### PR TITLE
[Merged by Bors] - feat: add lemma List.prod_map_one

### PR DESCRIPTION
**Pull request by BoltonBailey**
Adds the lemma, and `to_additive` tags it

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by YaelDillies** [comment]
I'm interested. What's your use case?

**Comment by BoltonBailey** [comment]
I have a project where I have lots of sums over `MvPolynomial` expressions. In the course of simplifying these expressions I sometimes come across this and want to simplify it.

I guess your next question will be "why not just use `simp`?. I think the main reason was probably for proof aesthetics - I conceptualize the simplification part of the proof as a sequence of steps where I first simplify various expressions involving sums over mapped lists, and then rearrange terms, and then I cast everything using `MvPolynomial.optionEquivRight` and then simplify further. I don't want to do this all at once because then it would be harder to understand as a human that each of these steps works and puts the hypothesis in a certain form. I therefore heavily use `simp only`, and I want to be able to reference this lemma directly, since everything else I am putting in the same `simp only` call.

This all got switched from `Finset` to `List` when I ported it from 3 to 4. Summing over a function seems to make sense for `Finset` because a `Finset` can't have duplicates, and so there is this pattern of summing over a mapped list that arises when I did this switch, and I think I wanted an analogous lemma to [Finset.sum_const_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Basic.html#Finset.sum_const_zero). Relatedly, we have things like [Multiset.prod_map_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Multiset/Basic.html#Multiset.prod_map_one). (Bit of a naming discrepancy there).

File: Mathlib/Data/List/BigOperators/Basic.lean [hunk 1]
```diff
@@ -143,6 +143,11 @@ theorem prod_map_mul {α : Type*} [CommMonoid α] {l : List ι} {f g : ι → α
 #align list.prod_map_mul List.prod_map_mul
 #align list.sum_map_add List.sum_map_add
 
+@[to_additive]
+theorem prod_map_one {α : Type*} [CommMonoid α] {l : List ι} :
+    (l.map fun _ => (1 : α)).prod = 1 := by
```
Comment (urkud) [inline Mathlib/Data/List/BigOperators/Basic.lean:148]: This should be true for a `MulOneClass`

Comment (BoltonBailey) [reply Mathlib/Data/List/BigOperators/Basic.lean:164]: Ok fair enough. The above assumptions all use `Monoid` so I have had to generalize there to fix this.

Comment (BoltonBailey) [reply Mathlib/Data/List/BigOperators/Basic.lean:164]: Yep, `prod_cons` does not necessarily hold for `MulOneClass` so I need a new `prod_one_cons` to cover this.

Comment (YaelDillies) [reply Mathlib/Data/List/BigOperators/Basic.lean:164]: Can you move it to the `MulOneClass` section?

File: Mathlib/Data/List/BigOperators/Basic.lean [hunk 2]
```diff
@@ -23,22 +23,45 @@ variable {ι α M N P M₀ G R : Type*}
...
+section Mul
 
-variable [Monoid M] [Monoid N] [Monoid P] {l l₁ l₂ : List M} {a : M}
+variable [Mul M] [One M]
 
 @[to_additive (attr := simp)]
 theorem prod_nil : ([] : List M).prod = 1 :=
   rfl
 #align list.prod_nil List.prod_nil
 #align list.sum_nil List.sum_nil
```
Comment (YaelDillies) [inline Mathlib/Data/List/BigOperators/Basic.lean:34]: I think there is not much point here, keep it assuming `MulOneClass`

Comment (BoltonBailey) [reply Mathlib/Data/List/BigOperators/Basic.lean]: Ok I have switched it to `MulOneClass` (initially it was `Monoid`)

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**